experimental GTK+4 support

rkitover opened the pull request:
Add configure support to detect GTK+4 (which is 3.90+) and set the
defines appropriately, including __WXGTK4__.

Use this define to compile out a few bits of code that are incompatible,
particularly some gdk functions, GTK+4 recommends using backend specific
functions for this, but I don't know how to do that, and this is just a
first pass to get things to compile.

rkitover commented:
I forgot to update the bakefile related stuff, sorry, will do that later today.

rkitover commented:
So I updated all the bakefile related stuff, and now there are some more compile errors to fix in the sources that got added.

I don't know why the xcode travis build is failing, it looks like it wants a program that is set as `SETFILE` in the makefiles by configure, but it's not being set for some reason, may not be related to this PR.

rkitover commented:
Ok so making the bakefiles work actually produces the dozens or hundreds of GTK API incompatibilities, which I am slowly fixing, can't promise that I will finish doing that though, may just give up.

vadz commented:
Thanks for looking at this! It's really disheartening to hear that there hundreds of API incompatibilities in GTK+ 4, I had a look at their [list of changes](https://developer.gnome.org/gtk4/stable/ch25s02.html) and while it seemed pretty bad, I didn't realize it was that bad... Unfortunately I haven't even started looking at GTK+ 4 yet, so I really can't do anything else than wishing you best of luck at this stage.

paulcor commented:
VZ wrote:
> Thanks for looking at this! It's really disheartening to hear that
> there hundreds of API incompatibilities in GTK+ 4

You may have forgotten that deprecation warnings are disabled by
default for the GTK+3 build (because there are so many it's hard to
spot other warnings). There are currently 289 warnings, representing 82
different deprecations. `GtkStock` is by far the largest problem, 138
warnings are due to that alone.

> Unfortunately I haven't even started looking at GTK+ 4 yet

I recently started to look at it. There are going to be some Problems.
Significant Problems.

vadz commented:
Damn :-( I really hoped they would calm down with API breaking changes as GTK+ matures more and more, but things seem to be going exactly in the opposite direction.

Of course, it could also be spun as an advertisement for wxWidgets: "Use our API and you won't have to rewrite your application every couple of years!". Provided that we can find resources to add GTK+ 4 support, that is...

paulcor commented:
Rafael Kitover wrote:
> Ok so making the bakefiles work actually produces the dozens or
> hundreds of GTK API incompatibilities, which I am slowly fixing,
> can't promise that I will finish doing that though, may just give up.

I think it would be better to submit separate fixes for individual
deprecation problems in the GTK+3 build. That will be much easier to
deal with than one giant patch changing many things at once, and can be
built and tested without needing GTK+3.9x.

rkitover commented:
@paulcor so how about I submit separate PRs for the following:

- autoconf support
- bakefile support
- each separate API fix/workaround/etc.

vadz commented:
FWIW I'm fine with using a single PR for this, splitting into independent commits is good enough IMHO. But the changes definitely should be split into separate commits, starting with the the separate fixes and with the build system changes coming at the end.

TIA!

rkitover committed:
GTK4: gtk_init_check() does not take args

In GTK+4 gtk_init_check() takes no arguments, while in GTK+3 you could
pass argc/argv.

I'm not quite sure how to pass any arguments to GTK+4, so just call
gtk_init_check() in GTK+4.

rkitover committed:
GTK4: different API for adding dialog button

Support the GTK+4 API for adding buttons to dialogs (via __WXGTK4__):

- set_icon_name() instead of image_new_from_stock()/set_image()

- gtk_box_pack_end() takes only two parameters

rkitover committed:
GTK4: no support for gtk_tree_view_set_rules_hint

Don't use gtk_tree_view_set_rules_hint() in GTK+4, it's not available.

It is also deprecated in GTK+3 since 3.14, and the documentation advises
to not use it at all, but I will leave the decision about whether to
remove it entirely to someone else.

rkitover committed:
GTK3/4: use widget scale instead of screen scale

GTK+4 does not support the gdk_screen_get_monitor_scale_factor() API
(since gdk has become more backend independent.)

Use gtk_widget_get_scale_factor() API instead on the app top window.

TODO: This does not solve the multi-monitor problem, for that it would
be necessary to use the scale factor of the widget on which the action
is being peformed.

rkitover commented:
@vadz @paulcor 

I've cleaned everything up and split it up into separate commits, can you guys take a look please?

vadz commented:
Thanks for the work! I'll have a more detailed look soon (unless Paul beats me to it), but Travis failures need to be fixed in any case.

vadz commented on src/unix/utilsx11.cpp:
```diff
@@ -2680,7 +2680,7 @@ bool
 {
 #ifdef __WXGTK__
-#if GTK_CHECK_VERSION(3,90,0)
+#ifdef __WXGTK3__
```
We can't remove the version check completely, this function is only available in 3.22.

We also presumably will need a run-time check if we use it with GTK+ 3 too.

paulcor replied:
`#ifdef __WXGTK4__`. `gtk_show_uri()` is fine for GTK+3.

vadz commented on include/wx/gtk/private/addremovectrl.h:
```diff
@@ -38,7 +38,9 @@ class wxAddRemoveImpl : public wxAddRemoveImplBase
         GtkStyleContext* context = gtk_widget_get_style_context(GTK_WIDGET(toolbar));
         gtk_style_context_add_class(context, GTK_STYLE_CLASS_INLINE_TOOLBAR);
+#ifndef __WXGTK4__
         gtk_style_context_set_junction_sides(context, GTK_JUNCTION_TOP);
```
Do things still look acceptably without this? Could you please check the "add/remove" dialog in the dialogs sample?

It's important to fix compilation, of course, but I'd like to avoid breaking the run-time appearance too...

rkitover replied:
I'll just drop this change for now and leave a comment, someone else can look at it when you get there, this is from a different compilation path.

vadz commented on build/aclocal/bakefile.m4:
```diff
@@ -745,7 +745,7 @@ AC_DEFUN([AC_BAKEFILE],
 
 
-    BAKEFILE_BAKEFILE_M4_VERSION="0.2.11"
+    BAKEFILE_BAKEFILE_M4_VERSION="0.2.10"
```
This doesn't look right.

vadz commented on build/bakefiles/common.bkl:
```diff
@@ -2,7 +2,7 @@
 <makefile>
 
-    <requires version="0.2.7"/>
+    <requires version="0.2.10"/>
```
Is this really necessary? Not that it hurts anyhow, of course, but I just wonder why was this changed.

rkitover replied:
The reason for the two changes above, is that there actually is no bakefile 0.2.11, the latest version of the legacy branch is 0.2.10, and the version check is broken so bakefile thinks that 0.2.10 is 0.2.1 and is below 0.2.7.

vadz replied:
This is indeed annoying and confusing (sorry...), but while the 0.2.11 release hasn't been done yet, the [latest branch](https://github.com/vslavik/bakefile/tree/legacy-0.2-branch) definitely does use 0.2.11.

rkitover replied:
Ah ok, I'll change that back then and use the git version.

vadz commented on configure.in:
```diff
@@ -6487,7 +6500,7 @@ if test "$wxUSE_UIACTIONSIMULATOR" = "yes" ; then
                 ],
                 [
-                    if test "$WXGTK3" = 1; then
+                    if test "$WXGTK3" = 1 -o "$WXGTK4" = 1; then
```
It might be more clear to define `WXGTK_3_OR_4` variable once and use it rather than testing for both of those in multiple places IMO.

paulcor replied:
`WXGTK4` implies `WXGTK3` here, so this change is not needed.

vadz commented:
Mostly this looks good, thanks for separating this in nicely atomic commits, this made it much simpler to review the changes.

And thanks again for the changes themselves, of course!

paulcor commented on configure.in:
```diff
@@ -5034,6 +5040,11 @@ else
+if test "$WXGTK4" = 1 ; then
+    AC_DEFINE_UNQUOTED(__WXGTK4__, 1)
+    WXGTK3=1
+    WXGTK2=1
```
Don't need `WXGTK2=1` here, the code immediately below will do it.

paulcor commented on configure.in:
```diff
@@ -6497,7 +6510,7 @@ if test "$wxUSE_UIACTIONSIMULATOR" = "yes" ; then
                 ]
             )
-        elif test "$WXGTK3" = 1; then
+        elif test "$WXGTK3" = 1 -o "$WXGTK4" = 1; then
```
`WXGTK4` implies `WXGTK3`

paulcor commented on configure.in:
```diff
@@ -7125,7 +7138,7 @@ if test "$wxUSE_WEBVIEW" = "yes"; then
     if test "$wxUSE_WEBVIEW_WEBKIT" = "yes"; then
         if test "$wxUSE_GTK" = 1; then
-            if test "$WXGTK3" = 1; then
+            if test "$WXGTK4" = 1 -o "$WXGTK3" = 1; then
```
`WXGTK4` implies `WXGTK3`

paulcor commented on configure.in:
```diff
@@ -7778,7 +7793,7 @@ elif test "$GXX" = yes ; then
     dnl toolkit versions and leaving this warning enabled prevents seeing any
     dnl other ones
-    if test "$WXGTK3" = 1 -o "$wxUSE_MAC" = 1 ; then
+    if test "$WXGTK4" = 1 -o "$WXGTK3" = 1 -o "$wxUSE_MAC" = 1 ; then
```
I don't think we want to disable deprecation warnings for GTK+4. At least not yet.

paulcor commented on configure.in:
```diff
@@ -7139,7 +7152,9 @@ if test "$wxUSE_WEBVIEW" = "yes"; then
-                if test "$WXGTK3" = 1; then
+                if test "$WXGTK4" = 1; then
+                    webkitgtk=webkitgtk-4.0
+                elif test "$WXGTK3" = 1; then
```
Could be simplified as `webkitgtk=webkitgtk-$TOOLKIT_VERSION.0`

paulcor commented on configure.in:
```diff
@@ -7977,7 +7992,9 @@ case "$TOOLKIT" in
-            if test "$WXGTK3" = 1; then
+            if test "$WXGTK4" = 1; then
+                TOOLKIT_DESC="$TOOLKIT_DESC 4"
+            elif test "$WXGTK3" = 1; then
```
Could simplify this as `TOOLKIT_DESC="$TOOLKIT_DESC $TOOLKIT_VERSION"`

paulcor commented on configure.in:
```diff
@@ -5489,7 +5500,9 @@ if test "$WXGTK2" = 1; then
-            if test "$WXGTK3" = 1; then
+            if test "$WXGTK4" = 1; then
+                gtk_unix_print="gtk+-unix-print-4.0"
+            elif test "$WXGTK3" = 1; then
```
I think this could be simplified as `gtk_unix_print="gtk+-unix-print-$TOOLKIT_VERSION.0"`

paulcor commented on src/gtk/assertdlg_gtk.cpp:
```diff
@@ -155,8 +156,8 @@ static void gtk_assert_dialog_save_backtrace_callback(GtkWidget*, GtkAssertDialo
-                                          GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
-                                          GTK_STOCK_SAVE, GTK_RESPONSE_ACCEPT,
+                                          wxGetTranslation("Cancel").mb_str(), GTK_RESPONSE_CANCEL,
+                                          wxGetTranslation("Save").mb_str(), GTK_RESPONSE_ACCEPT,
```
Should be `utf8_str()`

paulcor commented on build/aclocal/gtk-4.0.m4:
```diff
@@ -0,0 +1,208 @@
+# Configure paths for GTK+
```
Where does `gtk-4.0.m4` come from?

rkitover replied:
I made a copy of gtk-3.0.m4 and changed some things.

rkitover committed:
GTK4: show_uri() is gone

Use show_uri_on_window() instead of show_uri() for GTK+4, show_uri() is
not available.

rkitover committed:
autoconf and bakefile support for GTK+4

Add autoconf and bakefile support for detecting and using GTK+4, similar
to how GTK+3 is detected and used.

GTK+3 is checked for first before GTK+4, this can be overridden in
`./configure` using `--with-gtk=4`.

With GTK+4 the `__WXGTK4__` `#define` is also set in `setup.h.in`.

This commit does not regenerate any files.

rkitover committed:
regen bakefile/autoconf files based on 3559d83bc4

Run `bakefile_gen` and `sh autogen.sh` to regenerate all bakefile and
autoconf generated files with GTK+4 support from 3559d83bc4.

rkitover commented:
@vadz @paulcor I've made the requested changes.

rkitover commented:
Oops, forgot the utf8_str() change, pushed.

rkitover committed:
GTK3/4: GTK_STOCK_CANCEL deprecated, use labels

GTK+3 deprecates GTK_STOCK_CANCEL and GTK_STOCK_SAVE etc. and GTK+4
removes them entirely. The API documentation recommends using the
strings "Cancel" and "Save" instead, use them (or the appropriate
translation.)

rkitover commented:
Travis build should be fine now, was a wxString + GTK2 varargs issue, it's just waiting to do the OS X build now.
